Miscellaneous performance fixes for Crossgen2

SrivastavaAnubhav opened the pull request:
**Overall, these changes speed up Crossgen2 compilation of CoreLib by ~34%.** Most of the improvement comes from fewer/smaller allocations => need to do less GC.

- Create comparison functions for Enum (Enum.CompareTo boxes since it takes an Object)
- In ModuleToken, use the existing MetadataReader instead of making a new one each time
- Use Dictionary instead of ImmutableDictionary in CoreRTNameMangler to reduce allocations
- In CopiedFieldRvaNode, only read as much of the section as we need to copy

Before changes:
- Total average compilation time: **9522.11 ms**
- Phase breakdown (average):
  - Loading time: 682.91 ms
  - Graph processing time: **4189.59 ms**
    - Added 211462 nodes to mark
    - Dependency analysis time: 622.58 
    - Wall clock JIT time: 2988.49 
    - Total JIT time: 23079.96 ms (sum of all threads)
    - 22680 methods JITed
  - Emitting time: **4646.94 ms**

After changes:
- Total average compilation time: **6245.38 ms**
- Phase breakdown (average):
  - Loading time: 697.13 ms
  - Graph processing time: **3670.02 ms**
    - Added 211462 nodes to mark stack
    - Dependency analysis time: 631.47 ms
    - Wall clock JIT time: 2483.27 ms
    - Total JIT time: 19068.50 ms (sum of all threads)
    - 22680 methods JITed
  - Emitting time: **1875.50 ms**

Commit:
Miscellaneous performance fixes.
Create comparison functions for Enum (Enum.CompareTo boxes since it takes an Object).
In ModuleToken, use the existing MetadataReader instead of making a new one each time.
Use Dictionary instead of ImmutableDictionary in CoreRTNameMangler to reduce allocations.
In CopiedFieldRvaNode, only read as much of the section as we need to copy.

Commit:
Fix indentation.

SrivastavaAnubhav commented:
cc @dotnet/crossgen-contrib

jkotas commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/FieldFixupSignature.cs:
```diff
@@ -56,7 +56,7 @@ public override void AppendMangledName(NameMangler nameMangler, Utf8StringBuilde
         {
             FieldFixupSignature otherNode = (FieldFixupSignature)other;
-            int result = _fixupKind.CompareTo(otherNode._fixupKind);
+            int result = _fixupKind.CompareEnum(otherNode._fixupKind);
```
```suggestion
            int result = ((int)_fixupKind).CompareTo((int)otherNode._fixupKind);
```

huoyaoyuan replied:
Why not just `_fixupKind - otherNode._fixupKind` ?

fadimounir replied:
No need to call `CompareTo`. Just do the subtraction here directly as @huoyaoyuan suggests (and typecast if necessary)

jkotas replied:
Note that the subtraction does not work to compare arbitrary integers. It only works if the integers are small. It is probably the case here, but then you need to think about all the time about whether you are dealing with small integers or not.

lpereira replied:
Might be an [opportunity](https://gcc.godbolt.org/z/SpP7Lq) for changes in the JIT here: `(a == b) ? 0 : -Convert.ToInt32(a > b)` should be branch-free (`xor eax, eax; cmp edi, esi; setg al; neg eax`), or `Convert.ToInt32(a > b) - Convert.ToInt32(a < b)` should be something like `xor eax, eax; xor ecx,ecx; cmp edi, esi; seg al; setl cl; sub eax, ecx`. [ This doesn't seem to happen on a cursory glance.](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGBLAHYYGAWVIAKQcOxoGUhsACUHLjzXEA7A2wMAvLoUMA/AwAMDENoYA+QyfIWGccgG5V3AL7vecoaPKSfjK+wkoq1GpqfABmDOI6tmGaDK7eUbHxDAA8CsrJzm4RkdzJpoVqXkUl/H4idIHSsvJhnFWRySwQAgBuMFAYAHQAKhAAkkJ0Egm5Tqxdvf3DYxNT2bnlPJXqNcIiKA3aTX4taTzJOnAKG57e9CGiAKwHwc3KrcUlWjr6dmaOcJ0en1BiNxhhJpkckprgxKpUgA==)

lpereira replied:
Might be an opportunity to improve the codegen for `int.CompareTo()`.  It's currently not branch-free.  For instance, these two variants that works with arbitrary integers could be used instead of the branchy version:

- `(a == b) ? 0 : -Convert.ToInt32(a > b)` should be branch-free (`xor eax, eax; cmp edi, esi; setg al; neg eax`)
- `Convert.ToInt32(a > b) - Convert.ToInt32(a < b)` should be something like `xor eax, eax; xor ecx,ecx; cmp edi, esi; seg al; setl cl; sub eax, ecx`

(Might also be able to use a conditional move, although it doesn't seem that necessary.)

jkotas commented on src/coreclr/src/tools/Common/System/EnumExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+namespace System
+{
+    /// <summary>
+    /// Enum.CompareTo takes an Object, so using that requires boxing the enum we are comparing to (which is slow
```
I do not think it makes sense to have these in the shared file.

Instead, I would just add casts to int to every place where this is used (like what I have suggested below).

MichalStrehovsky commented on src/coreclr/src/tools/Common/Compiler/CoreRTNameMangler.cs:
```diff
@@ -166,11 +165,14 @@ private string DisambiguateName(string origName, ISet<string> set)
-            string mangledName;
-            if (_mangledTypeNames.TryGetValue(type, out mangledName))
-                return mangledName;
+            lock (this)
```
Can we break down the locking hierarchy so that there's not a single giant lock for type names, method names, field names, and mangled strings, but instead there's a lock for each?

In general, I'm worried that we're probably not using this from multithreaded code at all right now, so this looks like an improvement, but the locks will become a bottleneck as soon as we start doing that.

fadimounir replied:
How about using the lock free concurrent dictionaries here, to avoid locking while performing the lookup?

SrivastavaAnubhav replied:
@MichalStrehovsky We are using this from multithreaded code at the moment. I tried using more granulated locking and it didn't seem to make a performance difference. @davidwrighton and I thought that, since there's no performance difference right now, it would be better to keep it in this form where it's clear to people reading it that there are no deadlocks.

@fadimounir That's an option, but then it gets a little hard to analyze because it's hard to reason about whether each of the GetOrAdd functions must only be run once and so has to be either locked or wrapped in a Lazy. This + the limited time left is why we didn't end up using that right now, but perhaps @davidwrighton can provide more direction.

fadimounir replied:
> because it's hard to reason about whether each of the GetOrAdd functions must only be run once 

What do you mean? We can always do a .TryGetValue outside of a lock, and .TryAdd inside a lock (`Dictionary` has the same APIs too so we can use it as well). Locking while retrieving a value is not ideal

SrivastavaAnubhav replied:
Oh I see what you mean now. Yes I could change it to do that, but I'm not sure whether we even have contention there.

davidwrighton replied:
When using Dictionary you cannot use the TryGetValue apis outside of a lock. This worked before with ImmutableDictionary as that api allows lock-free reads. My expectation is that ConcurrentDictionary would be better in performance here than Dictionary, but I don't want to see us use that without seeing the performance impact in a profiler. I think we should check Anubhav's work in here as written, and explore finding even more performance in the future.

MichalStrehovsky commented on src/coreclr/src/tools/Common/System/EnumExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+            return (int)first - (int)second;
+        }
+
+        public static int CompareEnum(this ImportThunk.Kind first, ImportThunk.Kind second)
```
If the type is not in Common, code using the type should not be in Common.

+1 to Jan's advice.

fadimounir commented on src/coreclr/src/tools/Common/Compiler/CoreRTNameMangler.cs:
```diff
@@ -390,11 +391,14 @@ public override Utf8String GetMangledMethodName(MethodDesc method)
+            lock (this)
+            {
+                Utf8String mangledName;
+                if (_unqualifiedMangledMethodNames.TryGetValue(method, out mangledName))
```
Same here... if we make _unqualifiedMangledMethodNames a lock-free dictionary, we can avoid locking while looking up. We only need locking while adding new entries

fadimounir commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/CopiedFieldRvaNode.cs:
```diff
@@ -76,18 +75,20 @@ private unsafe byte[] GetRvaData(int targetPointerSize)
                     // We need to handle overlapping fields by reusing blobs based on the rva, and just update
                     // the size and contents
                     size = currentSize;
                     result = new byte[AlignmentHelper.AlignUp(size, targetPointerSize)];
```
We can also move the array allocation outside of the loop. No need to allocate and resize now

fadimounir commented:
The new perf numbers look pretty nice! Thank you!

Commit:
Address some PR comments.
Remove EnumExtensions in favor of casting to int and using CompareTo.
Move array allocation out of loop in CopiedFieldRvaNode.

davidwrighton commented on src/coreclr/src/tools/Common/System/EnumExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
```
I can see that you removed usage of these apis, but this file seems to still be here. Please remove this file.

SrivastavaAnubhav replied:
Oh oops, yes.

Commit:
Remove EnumExtensions file.

fadimounir commented:
@SrivastavaAnubhav, these changes are breaking determinism. See: https://dev.azure.com/dnceng/public/_build/results?buildId=454679&view=ms.vss-test-web.build-test-results-tab&runId=14512252&resultId=102657&paneView=attachments

Commit:
Fix determinism issue. Fix map file flag passing.
